File: clients/dart/test/rx_test.dart
import 'dart:async';
import 'dart:io';

import 'package:drift/native.dart';
import 'package:opto_sync_client/opto_sync_client.dart';
import 'package:opto_sync_client/rx.dart' as rx;
import 'package:rxdart/rxdart.dart';
import 'package:test/test.dart';

String locateCoreLibrary() {
  final env = Platform.environment['SYNCER_LIB_PATH'];
  if (env != null && env.isNotEmpty) return env;
  final starts = <String>[
    Directory.current.absolute.path,
    File.fromUri(Platform.script).parent.absolute.path,
  ];
  for (final start in starts) {
    var dir = Directory(start);
    for (var i = 0; i < 10; i++) {
      final buildDir =
          '${dir.path}${Platform.pathSeparator}syncer.c${Platform.pathSeparator}core${Platform.pathSeparator}build';
      if (Directory(buildDir).existsSync()) {
        final path = resolveSyncerLibraryPath(directory: buildDir);
        if (File(path).existsSync()) return path;
      }
      final parent = dir.parent;
      if (parent.path == dir.path) break;
      dir = parent;
    }
  }
  throw StateError('could not locate the syncer core library');
}

class _FakeLoop implements rx.SyncKicker {
  _FakeLoop(this.client, {this.acknowledge = true, this.fail = false});

  final OptoSyncClient client;
  final bool acknowledge;
  final bool fail;
  int hints = 0;
  int cycles = 0;

  @override
  void hint() => hints++;

  @override
  Future<ProtocolSyncCycleResult> syncNow() async {
    cycles++;
    if (fail) throw const SyncTransportException('sync failed');
    if (acknowledge) {
      final pending = await client.pendingMutations(limit: 1000);
      for (final row in pending) {
        await client.confirmSyncedUpTo(row.mutationId!, clientId: row.clientId);
      }
    }
    return const ProtocolSyncCycleResult(
      pushedMutations: 0,
      acknowledgedMutations: 0,
      pulledChanges: 0,
      installedSnapshots: 0,
      checkpoint: '0',
      hasMorePending: false,
    );
  }
}

void main() {
  late OptoSyncDatabase db;
  late FfiSyncer syncer;
  late OptoSyncClient client;

  setUpAll(() {
    syncer = FfiSyncer(libraryPath: locateCoreLibrary());
  });

  setUp(() {
    db = OptoSyncDatabase(NativeDatabase.memory());
    client = OptoSyncClient(
      db: db,
      syncer: syncer,
      overlaySyncer: syncer.overlay(),
    );
  });

  tearDown(() async {
    await db.close();
  });

  test('watchLocalView emits the optimistic view immediately after a write',
      () async {
    final authoritative = BehaviorSubject<Map<String, dynamic>?>.seeded({
      'id': 'r1',
      'title': 'server',
      'updatedAt': '100',
    });
    final emissions = <Map<String, dynamic>?>[];
    final sub = rx
        .watchLocalView(
          client: client,
          tableName: 'docs',
          recordId: 'r1',
          authoritative: authoritative,
        )
        .listen(emissions.add);

    await Future<void>.delayed(const Duration(milliseconds: 100));
    expect(emissions, hasLength(1));
    expect(emissions.single?['title'], 'server');

    await client.queueMutation('docs', 'r1', {
      'title': 'local edit',
      'updatedAt': '200',
    });
    await Future<void>.delayed(const Duration(milliseconds: 200));
    expect(emissions.last?['title'], 'local edit');

    await sub.cancel();
    await authoritative.close();
  });

  test('watchLocalView deduplicates canonically identical states', () async {
    final authoritative = BehaviorSubject<Map<String, dynamic>?>.seeded({
      'id': 'r2',
      'n': 1,
    });
    final emissions = <Map<String, dynamic>?>[];
    final sub = rx
        .watchLocalView(
          client: client,
          tableName: 'docs',
          recordId: 'r2',
          authoritative: authoritative,
        )
        .listen(emissions.add);
    await Future<void>.delayed(const Duration(milliseconds: 100));

    // Same canonical state, different key order and object identity.
    authoritative.add({'n': 1, 'id': 'r2'});
    await Future<void>.delayed(const Duration(milliseconds: 100));
    expect(emissions, hasLength(1));

    authoritative.add({'id': 'r2', 'n': 2});
    await Future<void>.delayed(const Duration(milliseconds: 100));
    expect(emissions, hasLength(2));

    await sub.cancel();
    await authoritative.close();
  });

  test('background optimism resolves once durably queued, no sync kicked',
      () async {
    final loop = _FakeLoop(client);
    final receipt = await rx.write(
      client,
      'docs',
      'r3',
      {'v': 1},
      optimism: rx.Optimism.background,
      loop: loop,
    );
    expect(receipt.optimism, rx.Optimism.background);
    expect(loop.cycles, 0);
    expect(loop.hints, 0);
    expect(await client.pendingMutations(), hasLength(1));
  });

  test('localFirst optimism queues then kicks a cycle without awaiting',
      () async {
    final loop = _FakeLoop(client);
    final receipt = await rx.write(client, 'docs', 'r4', {'v': 1}, loop: loop);
    expect(receipt.optimism, rx.Optimism.localFirst);
    expect(loop.hints, 1);
    expect(loop.cycles, 0);
  });

  test('awaitServer optimism resolves only after acknowledgement', () async {
    final loop = _FakeLoop(client);
    await rx.write(
      client,
      'docs',
      'r5',
      {'v': 1},
      optimism: rx.Optimism.awaitServer,
      loop: loop,
    );
    expect(loop.cycles, greaterThanOrEqualTo(1));
    expect(await client.pendingMutations(), isEmpty);
  });

  test('awaitServer rejects when the server never acknowledges — data stays queued',
      () async {
    final loop = _FakeLoop(client, acknowledge: false);
    await expectLater(
      rx.write(
        client,
        'docs',
        'r6',
        {'v': 1},
        optimism: rx.Optimism.awaitServer,
        loop: loop,
      ),
      throwsStateError,
    );
    expect(await client.pendingMutations(), hasLength(1));
  });

  test('awaitServer without a loop is a usage error', () async {
    await expectLater(
      rx.write(client, 'docs', 'r7', {'v': 1}, optimism: rx.Optimism.awaitServer),
      throwsArgumentError,
    );
  });

  test('writeDelete follows the same optimism contract', () async {
    final loop = _FakeLoop(client);
    await rx.writeDelete(
      client,
      'docs',
      'r8',
      optimism: rx.Optimism.awaitServer,
      loop: loop,
    );
    expect(await client.pendingMutations(), isEmpty);
  });

  test('hasUnsyncedWorkStream tracks the pending queue', () async {
    final states = <bool>[];
    final sub = rx.hasUnsyncedWorkStream(client).listen(states.add);
    await Future<void>.delayed(const Duration(milliseconds: 100));
    await client.queueMutation('docs', 'r9', {'v': 1});
    await Future<void>.delayed(const Duration(milliseconds: 200));
    expect(states, [false, true]);
    await sub.cancel();
  });
}

File: clients/ts/test/rx.test.mjs
import 'fake-indexeddb/auto';
import assert from 'node:assert/strict';
import test from 'node:test';
import { firstValueFrom } from 'rxjs';
import { filter, take, toArray } from 'rxjs';

import {
  OptoSyncClient,
  SYNC_STATUS,
  rx,
} from '../dist/index.js';

const { watchLocalView, write, writeDelete, hasUnsyncedWork$, createRxSyncLoop } = rx;
import { BehaviorSubject } from 'rxjs';

let databaseSequence = 0;
function makeClient() {
  databaseSequence += 1;
  return new OptoSyncClient({ databaseName: `rx-test-${databaseSequence}` });
}

function fakeLoop(client, behavior = {}) {
  return {
    hints: 0,
    cycles: 0,
    hint() {
      this.hints += 1;
    },
    async syncNow() {
      this.cycles += 1;
      if (behavior.fail) throw new Error('sync failed');
      if (behavior.acknowledge !== false) {
        // Pretend the server accepted everything queued so far.
        const pending = await client.pendingMutations();
        for (const row of pending) {
          await client.markMutation(row.id, SYNC_STATUS.SYNCED);
        }
      }
      return { pushedMutations: 0, acknowledgedMutations: 0, pulledChanges: 0, installedSnapshots: 0, checkpoint: '0', hasMorePending: false };
    },
  };
}

test('watchLocalView emits the optimistic view immediately after a queue write', async () => {
  const client = makeClient();
  const authoritative$ = new BehaviorSubject({ id: 'r1', title: 'server', updatedAt: '100' });

  const view$ = watchLocalView({
    client,
    tableName: 'docs',
    recordId: 'r1',
    authoritative$,
  });

  const emissions = [];
  const subscription = view$.subscribe((view) => emissions.push(view));
  await new Promise((resolve) => setTimeout(resolve, 50));
  assert.equal(emissions.length, 1);
  assert.equal(emissions[0].title, 'server');

  await client.queueMutation('docs', 'r1', { title: 'local edit', updatedAt: '200' });
  await new Promise((resolve) => setTimeout(resolve, 100));
  const last = emissions[emissions.length - 1];
  assert.equal(last.title, 'local edit');
  subscription.unsubscribe();
});

test('watchLocalView deduplicates canonically-identical states', async () => {
  const client = makeClient();
  const authoritative$ = new BehaviorSubject({ id: 'r2', n: 1 });
  const view$ = watchLocalView({
    client,
    tableName: 'docs',
    recordId: 'r2',
    authoritative$,
  });
  const emissions = [];
  const subscription = view$.subscribe((view) => emissions.push(view));
  await new Promise((resolve) => setTimeout(resolve, 50));

  // Server echoes a byte-identical state (fresh object): must not re-emit.
  authoritative$.next({ id: 'r2', n: 1 });
  await new Promise((resolve) => setTimeout(resolve, 50));
  assert.equal(emissions.length, 1);

  authoritative$.next({ id: 'r2', n: 2 });
  await new Promise((resolve) => setTimeout(resolve, 50));
  assert.equal(emissions.length, 2);
  subscription.unsubscribe();
});

test("write optimism 'background' resolves once durably queued and never syncs", async () => {
  const client = makeClient();
  const loop = fakeLoop(client);
  const receipt = await write(client, 'docs', 'r3', { v: 1 }, { optimism: 'background', loop });
  assert.equal(receipt.optimism, 'background');
  const row = await client.db.localMutations.get(receipt.queuedMutationId);
  assert.equal(row.syncStatus, SYNC_STATUS.PENDING);
  assert.equal(loop.cycles, 0);
  assert.equal(loop.hints, 0);
});

test("write optimism 'local-first' queues then kicks a cycle without awaiting it", async () => {
  const client = makeClient();
  const loop = fakeLoop(client);
  const receipt = await write(client, 'docs', 'r4', { v: 1 }, { loop });
  assert.equal(receipt.optimism, 'local-first');
  assert.equal(loop.hints, 1);
  assert.equal(loop.cycles, 0);
});

test("write optimism 'await-server' resolves only after acknowledgement", async () => {
  const client = makeClient();
  const loop = fakeLoop(client);
  const receipt = await write(client, 'docs', 'r5', { v: 1 }, { optimism: 'await-server', loop });
  assert.ok(loop.cycles >= 1);
  const row = await client.db.localMutations.get(receipt.queuedMutationId);
  assert.equal(row.syncStatus, SYNC_STATUS.SYNCED);
});

test("write optimism 'await-server' rejects when the server never acknowledges", async () => {
  const client = makeClient();
  const loop = fakeLoop(client, { acknowledge: false });
  await assert.rejects(
    write(client, 'docs', 'r6', { v: 1 }, { optimism: 'await-server', loop }),
    /did not acknowledge/,
  );
  // The local write is still durably queued — optimism never loses data.
  const pending = await client.pendingMutations('docs');
  assert.equal(pending.length, 1);
});

test("write optimism 'await-server' without a loop is a usage error", async () => {
  const client = makeClient();
  await assert.rejects(
    write(client, 'docs', 'r7', { v: 1 }, { optimism: 'await-server' }),
    RangeError,
  );
});

test('writeDelete follows the same optimism contract', async () => {
  const client = makeClient();
  const loop = fakeLoop(client);
  const receipt = await writeDelete(client, 'docs', 'r8', { optimism: 'await-server', loop });
  const row = await client.db.localMutations.get(receipt.queuedMutationId);
  assert.equal(row.operation, 'delete');
  assert.equal(row.syncStatus, SYNC_STATUS.SYNCED);
});

test('hasUnsyncedWork$ tracks the pending queue', async () => {
  const client = makeClient();
  const states = [];
  const subscription = hasUnsyncedWork$(client).subscribe((v) => states.push(v));
  await new Promise((resolve) => setTimeout(resolve, 50));
  await client.queueMutation('docs', 'r9', { v: 1 });
  await new Promise((resolve) => setTimeout(resolve, 100));
  assert.deepEqual(states, [false, true]);
  subscription.unsubscribe();
});

test('createRxSyncLoop streams state transitions with replay for late subscribers', async () => {
  const client = makeClient();
  const transport = {
    push: async () => ({ protocolVersion: 1, lastMutationId: '0', results: [] }),
    pull: async (checkpoint) => ({ protocolVersion: 1, checkpoint, hasMore: false, changes: [] }),
    snapshot: async () => ({ protocolVersion: 1, checkpoint: '0', records: [] }),
  };
  const { ProtocolSyncLoop } = await import('../dist/index.js');
  const { loop, state$ } = createRxSyncLoop(
    (onStateChange) =>
      new ProtocolSyncLoop(
        client,
        transport,
        {
          applyChanges: async () => undefined,
          replaceAuthoritative: async () => undefined,
        },
        { onStateChange, observeBrowserLifecycle: false },
      ),
  );
  // An un-started loop settles to 'stopped' after a one-shot cycle.
  const statuses = firstValueFrom(
    state$.pipe(
      filter((state) => state.status === 'stopped'),
      take(1),
    ),
  );
  await loop.syncNow();
  assert.equal((await statuses).status, 'stopped');
  // Late subscriber sees the latest state via replay.
  const replayed = await firstValueFrom(state$.pipe(take(1)));
  assert.equal(replayed.status, 'stopped');
});
